curvefs/src/mds/space/volume_space.cpp

```diff
@@ -487,6 +650,235 @@ SpaceErrCode VolumeSpace::ExtendVolume() {
+
+            summary_[offset] -= lastDeallocatableSize;
+            if (summary_[offset] == 0) {
+                summary_.erase(offset);
+                LOG(INFO) << "VolumeSpace remove block group from summary, no "
+                             "need deallocatable, "
+                             "fsId="
+                          << fsId_ << ", blcokGroupOffset=" << offset;
+            }
+
+            group.set_deallocatablesize(reportGroups[offset]);
+        }
+    }
+}
+
+void VolumeSpace::UpdateDeallocatingBlockGroup(
+    uint32_t metaserverId, const BlockGroupDeallcateStatusMap &stats) {
+    LockGuard lk(mtx_);
+
+    VLOG(6) << "VolumeSpace update deallocating block group from metaserver="
+            << metaserverId << ", fsId=" << fsId_
+            << ", stats size=" << stats.size();
+
+    // get completed deallocate blockgroup
+    std::vector<uint64_t> doneGroups;
+    for (auto &stat : stats) {
+        auto offset = stat.first;
+        auto status = stat.second;
+
+        auto iter = deallocatingGroups_.find(offset);
+        if (iter == deallocatingGroups_.end()) {
+            LOG(ERROR) << "VolumeSpace block group not found in "
+                          "deallocatingGroups_, fsId="
+                       << fsId_  << ", blockGroupOffset=" << offset;
+            continue;
+        }
+
+        VLOG(6) << "VolumeSpace get block group stat from metaserver="
+                << metaserverId << ", fsId=" << fsId_
+                << ", blockGroupOffset=" << offset
+                << ", status=" << BlockGroupDeallcateStatusCode_Name(status);
+
+        auto alreadyDeallocated =
+            std::find(iter->second.deallocated().begin(),
+                      iter->second.deallocated().end(), metaserverId);
+        if (status == BlockGroupDeallcateStatusCode::BGDP_DONE &&
+            alreadyDeallocated == iter->second.deallocated().end()) {
+            doneGroups.emplace_back(offset);
+            LOG(INFO) << "VolumeSpace block group is deallocated done, fsId: "
+                      << fsId_ << ", blcokGroupOffset: " << offset
+                      << ", metaserverId: " << metaserverId;
+        }
+    }
+
+    // update the metaserver from the deallocating state of the blockgroup to
+    // the deallocated state
+    for (auto offset : doneGroups) {
+        auto iter = deallocatingGroups_.find(offset);
+        assert(iter != deallocatingGroups_.end());
+
+        std::vector<uint32_t> newDeallocating;
+        for (auto id : iter->second.deallocating()) {
+            if (metaserverId == id) {
+                continue;
+            }
+            newDeallocating.emplace_back(id);
+        }
+
+        iter->second.add_deallocated(metaserverId);
+        auto mutableDeallocating = iter->second.mutable_deallocating();
+        mutableDeallocating->Resize(newDeallocating.size(), 0);
+        std::copy(newDeallocating.begin(), newDeallocating.end(),
+                  mutableDeallocating->begin());
```

Comment:
consider use erase-remove like you have done in other codes

Reply:
fix